lib/backend/syncersv1/updateprocessors/globalnetworkpolicyprocessor_test.go

```diff
@@ -22,224 +22,151 @@ import (
 	"github.com/projectcalico/libcalico-go/lib/backend/model"
 	"github.com/projectcalico/libcalico-go/lib/backend/syncersv1/updateprocessors"
 	cnet "github.com/projectcalico/libcalico-go/lib/net"
-	"github.com/projectcalico/libcalico-go/lib/numorstring"
 )
 
 var _ = Describe("Test the GlobalNetworkPolicy update processor", func() {
-	name1 := "name1"
-	name2 := "name2"
-
-	v3GlobalNetworkPolicyKey1 := model.ResourceKey{
-		Kind: apiv3.KindGlobalNetworkPolicy,
-		Name: name1,
-	}
-	v3GlobalNetworkPolicyKey2 := model.ResourceKey{
-		Kind: apiv3.KindGlobalNetworkPolicy,
-		Name: name2,
-	}
-	v1GlobalNetworkPolicyKey1 := model.PolicyKey{
-		Name: name1,
-	}
-	v1GlobalNetworkPolicyKey2 := model.PolicyKey{
-		Name: name2,
-	}
-
-	It("should handle conversion of valid GlobalNetworkPolicys", func() {
+	emptyGNPKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "empty"}
+	emptyGNP := apiv3.NewGlobalNetworkPolicy()
+
+	minimalGNPKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "minimal"}
+	minimalGNP := apiv3.NewGlobalNetworkPolicy()
+	minimalGNP.Spec.PreDNAT = true
+	minimalGNP.Spec.ApplyOnForward = true
+
+	fullGNPKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "simple"}
+	fullGNP := apiv3.NewGlobalNetworkPolicy()
+	fullGNP.Namespace = "default"
+	fullGNP.Spec.Order = &Order
+	fullGNP.Spec.Ingress = []apiv3.Rule{TestIngressRule}
+	fullGNP.Spec.Egress = []apiv3.Rule{TestEgressRule}
+	fullGNP.Spec.Selector = "mylabel == selectme"
+	fullGNP.Spec.DoNotTrack = true
+	fullGNP.Spec.PreDNAT = false
+	fullGNP.Spec.ApplyOnForward = true
+	fullGNP.Spec.Types = []apiv3.PolicyType{apiv3.PolicyTypeIngress}
+
+	// GlobalNetworkPolicies with valid and invalid ServiceAccountSelectors
+	validSASelectorKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "validSASelector"}
+	validSASelector := fullGNP.DeepCopy()
+	validSASelector.Spec.ServiceAccountSelector = "role == 'development'"
+
+	invalidSASelectorKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "invalidSASelector"}
+	invalidSASelector := fullGNP.DeepCopy()
+	invalidSASelector.Spec.ServiceAccountSelector = "role 'development'"
+
+	// GlobalNetworkPolicies with valid and invalid NamespaceSelectors
+	validNSSelectorKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "validNSSelector"}
+	validNSSelector := fullGNP.DeepCopy()
+	validNSSelector.Spec.NamespaceSelector = "name == 'testing'"
+
+	invalidNSSelectorKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "invalidNSSelector"}
+	invalidNSSelector := fullGNP.DeepCopy()
+	invalidNSSelector.Spec.NamespaceSelector = "name 'testing'"
+
+	SAandNSSelectorKey := model.ResourceKey{Kind: apiv3.KindGlobalNetworkPolicy, Name: "SAandNSSelector"}
+	SAandNSSelector := fullGNP.DeepCopy()
+	SAandNSSelector.Spec.ServiceAccountSelector = "role == 'development'"
+	SAandNSSelector.Spec.NamespaceSelector = "name == 'testing'"
```

Comment:
If I'm reading correctly, I think we have cases for the following valid combinations now:

- Selector
- Selector, SA selector
- Selector, NS selector
- Selector, NS selector, SA selector.

I think we should have a few more, though. 

Namely: 
- test cases where `Spec.Selector` is empty but others aren't in various combinations. 
- test cases which check the "all()" behavior is correct for both the NS and SA selectors.


It would be really great if in addition we could actually run a selection against some test label sets and make sure the rendered selector actually selects the things we expect, and doesn't select the things we don't want it to.

e.g., using this: https://github.com/projectcalico/libcalico-go/blob/master/lib/selector/selector.go#L21-L22

Reply:
Added the new test cases with an empty `Spec.Selector`. The tests against actual labels would take some time, not sure we wanna do those in this sprint tho.